File: examples/vb.net/USB2LIN/USB2XXX_LINEX_Test/USB2XXX_LIN_Test/USB2XXX_LIN_Test/USB2XXX_LIN_Test.vb

#Const LIN_MASTER = True

Module USB2XXX_LIN_Test
    '若运行程序提示找不到USB2XXX.dll文件，请将USB2XXX.dll文件和libusb-1.0.dll文件拷贝到exe程序输出目录下，比如bin/Debug目录下
    Sub Main()
        Dim ret As Int32
        Dim State As Boolean
        Dim DeviceHandle(20) As UInt32
        Dim LINIndex As Byte
        LINIndex = 0
        '扫描当前连接到电脑的LIN适配器
        ret = USB_ScanDevice(DeviceHandle)
        If ret <= 0 Then
            Console.WriteLine("No device connected")
        End If
        '打开设备
        State = USB_OpenDevice(DeviceHandle(0))
        If State Then
            Console.WriteLine("Open device success!")
        Else
            Console.WriteLine("Open device failed!")
            Return
        End If
        '初始化配置LIN
#If LIN_MASTER Then
        ret = LIN_EX_Init(DeviceHandle(0), LINIndex, 19200, LIN_EX_MASTER)
#Else
        ret = LIN_EX_Init(DeviceHandle(0), LINIndex, 19200, LIN_EX_SLAVE)
#End If
        If ret <> LIN_EX_SUCCESS Then
            Console.WriteLine("Config LIN failed!")
            Return
        Else
            Console.WriteLine("Config LIN success!")
        End If

#If LIN_MASTER Then
        '主机循环写数据
        Dim ID As Byte
        For ID = 0 To 9
            Dim msg_len As Byte = 1
            Dim msg_in(msg_len - 1) As LIN_EX_MSG
            Dim msg_out(msg_len - 1) As LIN_EX_MSG
            Dim i As Byte
            ReDim msg_in(0)
            ReDim msg_in(0).Data(8)
            ReDim msg_out(0)
            ReDim msg_out(0).Data(8)
            For i = 0 To 7
                msg_in(0).Data(i) = ID + i  '具体数据根据实际情况修改
            Next
            msg_in(0).DataLen = 8   '指定需要发送的数据字节数
            msg_in(0).PID = ID      '传入ID即可，底层会自动计算校验位
            msg_in(0).CheckType = LIN_EX_CHECK_EXT  '校验类型
            msg_in(0).MsgType = LIN_EX_MSG_TYPE_MW  '消息类型，主机写数据
            msg_in(0).Timestamp = 0                 '数据发送完毕后不需要延时
            ret = LIN_EX_MasterSync(DeviceHandle(0), LINIndex, msg_in, msg_out, msg_len)
            If ret <> msg_len Then
                Console.WriteLine("LIN write data failed!")
                Return
            Else
                '通过对输出的消息进行判断可以知道发送出去的数据是否成功发送
                Console.Write("Master LIN Write: PID[ID] = 0x{0:X2}[0x{1:X2}] Data = ", msg_out(0).PID, msg_out(0).PID And &H3F)
                If msg_out(0).DataLen > 0 Then
                    For i = 0 To msg_out(0).DataLen - 1
                        Console.Write("0x{0:X2} ", msg_out(0).Data(i))
                    Next i
                End If
                Console.WriteLine("")
            End If
            '延时
            Threading.Thread.Sleep(10)
        Next ID
        '主机循环读数据
        For ID = 10 To 15
            Dim msg_len As Byte = 1
            Dim msg_in(msg_len - 1) As LIN_EX_MSG
            Dim msg_out(msg_len - 1) As LIN_EX_MSG
            Dim i As Byte
            ReDim msg_in(0)
            ReDim msg_in(0).Data(8)
            ReDim msg_out(0)
            ReDim msg_out(0).Data(8)
            msg_in(0).PID = ID  '传入ID即可，底层会自动计算校验位
            msg_in(0).MsgType = LIN_EX_MSG_TYPE_MW  '消息类型，主机读数据
            ReDim msg_in(0).Data(8)
            ret = LIN_EX_MasterSync(DeviceHandle(0), LINIndex, msg_in, msg_out, msg_len)
            If ret <> msg_len Then
                Console.WriteLine("LIN read data failed!")
                Return
            Else
                '主机读到的数据存放在数据消息中
                Console.Write("Master LIN Read : PID[ID] = 0x{0:X2}[0x{1:X2}] Data = ", msg_out(0).PID, msg_out(0).PID And &H3F)
                If msg_out(0).DataLen > 0 Then  '具体读到了多少字节数据，可以通过DataLen参数获取
                    For i = 0 To msg_out(0).DataLen - 1
                        Console.Write("0x{0:X2} ", msg_out(0).Data(i))
                    Next i
                End If
                Console.WriteLine("")
            End If
            '延时
            Threading.Thread.Sleep(10)
        Next ID
#Else
        '设置从机模式下所有ID都为从接收数据模式，这样就可以获取到主机发送过来的所有数据,初始化配置为从机后，默认所有ID都为接收数据模式，所以若是监听LIN总线数据，这个函数可以不用调用
        Dim LINSlaveData(64) As LIN_EX_MSG
        Dim i As Int32
        For i = 0 To 63
            ReDim LINSlaveData(i)
            LINSlaveData(i).PID = i      'ID值
            LINSlaveData(i).MsgType = LIN_EX_MSG_TYPE_SR 'LIN_EX_MSG_TYPE_SR：从机接收数据模式，配置为从机后所有ID默认为该模式
            ReDim LINSlaveData(i).Data(8)
        Next i
        ret = LIN_EX_SlaveSetIDMode(DeviceHandle(0), LINIndex, LINSlaveData, 64)
        If ret <> LIN_EX_SUCCESS Then
            Console.WriteLine("Set LIN operation mode failed!")
            Return
        Else
            Console.WriteLine("Set LIN operation mode success!")
        End If
        '从机接收数据,若要连续不断的监控LIN总线数据，可以循环调用LIN_SlaveGetData函数
        Dim LINSlaveDataBuffer(1024) As LIN_EX_MSG '为了防止缓冲区溢出，可以将接收数据缓冲区设置大一点
        For i = 0 To LINSlaveDataBuffer.Length - 1
            ReDim LINSlaveDataBuffer(i)
            ReDim LINSlaveDataBuffer(i).Data(8)
        Next
        Console.WriteLine("LIN slave start get data...")
        While True
            ret = LIN_EX_SlaveGetData(DeviceHandle(0), LINIndex, LINSlaveDataBuffer)
            If ret < LIN_EX_SUCCESS Then
                Console.WriteLine("LIN slave read data error!")
                Return
            ElseIf ret > 0 Then
                For i = 0 To ret - 1
                    Console.Write("Slave LIN Read : PID[ID] = 0x{0:X2}[0x{1:X2}] Data = ", LINSlaveDataBuffer(i).PID, LINSlaveDataBuffer(i).PID And &H3F)
                    If LINSlaveDataBuffer(i).DataLen > 0 Then
                        Dim j As Int32
                        For j = 0 To LINSlaveDataBuffer(i).DataLen - 1
                            Console.Write("0x{0:X2} ", LINSlaveDataBuffer(i).Data(j))
                        Next j
                    End If
                    Console.WriteLine("")
                Next i
            Else
                'Console.WriteLine("LIN slave read no data!")
                '延时
                Threading.Thread.Sleep(100)
            End If
        End While
#End If

    End Sub

End Module

File: examples/C&C++/USB2LIN/USB2XXXLINExTest/source/USB2XXXLINTest.cpp
  /*
  ******************************************************************************
  * @file     : USB2XXXLINTest.cpp
  * @Copyright: usbxyz 
  * @Revision : ver 1.0
  * @Date     : 2014/12/19 9:33
  * @brief    : USB2XXX LIN test demo
  ******************************************************************************
  * @attention
  *
  * Copyright 2009-2014, usbxyz.com
  * http://www.usbxyz.com/
  * All Rights Reserved
  * 
  ******************************************************************************
  */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "usb_device.h"
#include "usb2lin_ex.h"

#define GET_FIRMWARE_INFO       1//获取固件信息
#define LIN_MASTER_TEST         0//主机模式读写测试
#define LIN_SLAVE_TEST          0//从机模式读写测试
#define LIN_MASTER_SCH_TEST     1//主机模式调度表测试

int main(int argc, const char* argv[])
{
#if GET_FIRMWARE_INFO
    DEVICE_INFO DevInfo;
#endif
    int DevHandle[10];
    int LINMasterIndex = 0;
    int LINSlaveIndex = 0;
    int DevIndex = 0;
    bool state;
    int ret;
    char *MSGTypeStr[]={"UN","MW","MR","SW","SR","BK","SY","ID","DT","CK"};
    char *CKTypeStr[]={"STD","EXT","USER","NONE","ERROR"};
    //扫描查找设备
    ret = USB_ScanDevice(DevHandle);
    if(ret <= 0){
        printf("No device connected!\n");
        return 0;
    }
    /*
    //当电脑上同时接有多个LIN适配器时，可以根据设备号指定对应的设备
    for(int i=0;i<ret;i++){
        if(DevHandle[i]==0x1C0){
            DevIndex = i;
            break;
        }
    }
    */
    //打开设备
    state = USB_OpenDevice(DevHandle[DevIndex]);
    if(!state){
        printf("Open device error!\n");
        return 0;
    }
#if GET_FIRMWARE_INFO
    char FunctionStr[256]={0};
    //获取固件信息
    state = DEV_GetDeviceInfo(DevHandle[DevIndex],&DevInfo,FunctionStr);
    if(!state){
        printf("Get device infomation error!\n");
        return 0;
    }else{
        printf("Firmware Info:\n");
	    printf("Firmware Name:%s\n",DevInfo.FirmwareName);
        printf("Firmware Build Date:%s\n",DevInfo.BuildDate);
        printf("Firmware Version:v%d.%d.%d\n",(DevInfo.FirmwareVersion>>24)&0xFF,(DevInfo.FirmwareVersion>>16)&0xFF,DevInfo.FirmwareVersion&0xFFFF);
        printf("Hardware Version:v%d.%d.%d\n",(DevInfo.HardwareVersion>>24)&0xFF,(DevInfo.HardwareVersion>>16)&0xFF,DevInfo.HardwareVersion&0xFFFF);
	    printf("Firmware Functions:%s\n",FunctionStr);
        printf("Firmware SerialNumber:%08X%08X%08X\n",DevInfo.SerialNumber[0],DevInfo.SerialNumber[1],DevInfo.SerialNumber[2]);
    }
#endif
//    USB_ResetDevice(DevHandle[DevIndex]);//复位设备测试函数
//    return 0;
#if LIN_MASTER_SCH_TEST||LIN_MASTER_TEST
    //初始化配置LIN
    ret = LIN_EX_Init(DevHandle[DevIndex],LINMasterIndex,9600,1);
    if(ret != LIN_EX_SUCCESS){
        printf("Config LIN failed!\n");
        return 0;
    }else{
        printf("Config LIN Success!\n");
    }
#endif
#if LIN_MASTER_TEST
    LIN_EX_MSG LINMsg[5];
    LIN_EX_MSG LINOutMsg[10];
    unsigned int MsgLen = 5;
    LINMsg[0].MsgType = LIN_EX_MSG_TYPE_BK;
    LINMsg[0].Timestamp = 10;//发送该帧数据之后的延时时间，最小建议设置为1
    //后面4个ID全部为主机发送数据模式
    for(int f=1;f<MsgLen;f++){
        LINMsg[f].MsgType = LIN_EX_MSG_TYPE_MW;//主机写数据模式
        LINMsg[f].DataLen = 8;
        for(int i=0;i<LINMsg[1].DataLen;i++){
            LINMsg[f].Data[i] = (f<<4)|i;
        }
        LINMsg[f].Timestamp = 10;//发送该帧数据之后的延时时间，最小建议设置为1
        LINMsg[f].CheckType = LIN_EX_CHECK_EXT;//设置好校验类型后，适配器会根据校验类型自动计算校验数据
        LINMsg[f].PID = f+1;
    }
    ret = LIN_EX_MasterSync(DevHandle[DevIndex],LINMasterIndex,LINMsg,LINOutMsg,MsgLen);
    if(ret < LIN_EX_SUCCESS){
        printf("MasterSync LIN failed!\n");
        return 0;
    }else{
        printf("MsgLen = %d\n",ret);
        for(int i=0;i<ret;i++){
            printf("%s SYNC[%02X] PID[%02X] ",MSGTypeStr[LINOutMsg[i].MsgType],LINOutMsg[i].Sync,LINOutMsg[i].PID);
            for(int j=0;j<LINOutMsg[i].DataLen;j++){
                printf("%02X ",LINOutMsg[i].Data[j]);
            }
            printf("[%s][%02X] [%02d:%02d:%02d.%03d]\n",CKTypeStr[LINOutMsg[i].CheckType],LINOutMsg[i].Check,(LINOutMsg[i].Timestamp/36000000)%60,(LINOutMsg[i].Timestamp/600000)%60,(LINOutMsg[i].Timestamp/10000)%60,(LINOutMsg[i].Timestamp/10)%1000);
        }
    }
    //后面4个ID全部为主机读数据模式
    for(int f=1;f<MsgLen;f++){
        LINMsg[f].MsgType = LIN_EX_MSG_TYPE_MR;//主机读数据模式
        LINMsg[f].Timestamp = 10;//主机读取完数据之后的延时时间，最小建议设置为1
        LINMsg[f].PID = f+1;
    }
    ret = LIN_EX_MasterSync(DevHandle[DevIndex],LINMasterIndex,LINMsg,LINOutMsg,MsgLen);
    if(ret < LIN_EX_SUCCESS){
        printf("MasterSync LIN failed!\n");
        return 0;
    }else{
        printf("MsgLen = %d\n",ret);
        for(int i=0;i<ret;i++){
            printf("%s SYNC[%02X] PID[%02X] ",MSGTypeStr[LINOutMsg[i].MsgType],LINOutMsg[i].Sync,LINOutMsg[i].PID);
            for(int j=0;j<LINOutMsg[i].DataLen;j++){
                printf("%02X ",LINOutMsg[i].Data[j]);
            }
            printf("[%s][%02X] [%02d:%02d:%02d.%03d]\n",CKTypeStr[LINOutMsg[i].CheckType],LINOutMsg[i].Check,(LINOutMsg[i].Timestamp/36000000)%60,(LINOutMsg[i].Timestamp/600000)%60,(LINOutMsg[i].Timestamp/10000)%60,(LINOutMsg[i].Timestamp/10)%1000);
        }
    }
#endif

#if LIN_SLAVE_TEST
    //初始化配置LIN
    ret = LIN_EX_Init(DevHandle[DevIndex],LINSlaveIndex,9600,0);
    if(ret != LIN_EX_SUCCESS){
        printf("Config LIN failed!\n");
        return 0;
    }else{
        printf("Config LIN Success!\n");
    }
    printf("Start Get LIN Data...\n");
    //设置ID为LIN_EX_MSG_TYPE_SW模式，这样主机就可以读取到数据
    LIN_EX_MSG LINSlaveMsg[10];
    for(int i=0;i<10;i++){
        LINSlaveMsg[i].PID = i;
        LINSlaveMsg[i].CheckType = LIN_EX_CHECK_EXT;
        LINSlaveMsg[i].DataLen = 7;
        for(int j=0;j<LINSlaveMsg[i].DataLen;j++){
            LINSlaveMsg[i].Data[j]=(i<<4)|j;
        }
        LINSlaveMsg[i].MsgType = LIN_EX_MSG_TYPE_SW;//从机发送数据模式
    }
    ret = LIN_EX_SlaveSetIDMode(DevHandle[DevIndex],LINSlaveIndex,LINSlaveMsg,10);
    if(ret != LIN_EX_SUCCESS){
        printf("Config LIN ID Mode failed!\n");
        return 0;
    }else{
        printf("Config LIN ID Mode Success!\n");
    }
    //循环获取接收到的数据，该操作可以用作LIN总线数据监控
    while(1)
    {
        LIN_EX_MSG LINMsg[1024];//缓冲区尽量大一点，防止益处
        int ret = LIN_EX_SlaveGetData(DevHandle[DevIndex],LINSlaveIndex,LINMsg);
        for(int i=0;i<ret;i++){
            printf("%s SYNC[%02X] PID[%02X] ",MSGTypeStr[LINMsg[i].MsgType],LINMsg[i].Sync,LINMsg[i].PID);
            for(int j=0;j<LINMsg[i].DataLen;j++){
                printf("%02X ",LINMsg[i].Data[j]);
            }
            printf("[%s][%02X] [%02d:%02d:%02d.%03d]\n",CKTypeStr[LINMsg[i].CheckType],LINMsg[i].Check,(LINMsg[i].Timestamp/36000000)%60,(LINMsg[i].Timestamp/600000)%60,(LINMsg[i].Timestamp/10000)%60,(LINMsg[i].Timestamp/10)%1000);
        }
        Sleep(1000);
    }
#endif
#if LIN_MASTER_SCH_TEST
    LIN_EX_MSG LINMsg[6];
    LIN_EX_MSG LINOutMsg[10];
    unsigned int MsgLen = 6;
    LINMsg[0].MsgType = LIN_EX_MSG_TYPE_BK;
    LINMsg[0].Timestamp = 20;//当前帧发送时，下一帧数据发送的时间间隔，单位为毫秒
    for(int f=1;f<MsgLen;f++){
        LINMsg[f].MsgType = LIN_EX_MSG_TYPE_MW;
        LINMsg[f].DataLen = 8;
        for(int i=0;i<LINMsg[1].DataLen;i++){
            LINMsg[f].Data[i] = (f<<4)|i;
        }
        LINMsg[f].Timestamp = 20+f*10;//当前帧发送时，下一帧数据发送的时间间隔，单位为毫秒
        LINMsg[f].CheckType = LIN_EX_CHECK_EXT;
        LINMsg[f].PID = f+1;
    }
    //ID=6，该帧为主机读数据模式，也就是主机发送帧头，从机返回数据
    LINMsg[5].MsgType = LIN_EX_MSG_TYPE_MR;
    LINMsg[5].PID = 6;
    LINMsg[5].Timestamp = 20;
    //启动调度表
    ret = LIN_EX_MasterStartSch(DevHandle[DevIndex],LINMasterIndex,LINMsg,MsgLen);
    if(ret < LIN_EX_SUCCESS){
        printf("Master schedule start failed!\n");
        return 0;
    }else{
        printf("Master schedule start success!\n");
    }
    //循环获取调度表数据
    for(int i=0;i<10;i++){
        Sleep(1000);
        ret = LIN_EX_MasterGetSch(DevHandle[DevIndex],LINMasterIndex,LINOutMsg);
        if(ret < LIN_EX_SUCCESS){
            printf("Master get schedule failed!\n");
            return 0;
        }else{
            printf("[0]MsgLen = %d\n",ret);
            for(int i=0;i<ret;i++){
                printf("[0]%s SYNC[%02X] PID[%02X] ",MSGTypeStr[LINOutMsg[i].MsgType],LINOutMsg[i].Sync,LINOutMsg[i].PID);
                for(int j=0;j<LINOutMsg[i].DataLen;j++){
                    printf("%02X ",LINOutMsg[i].Data[j]);
                }
                printf("\n");
            }
        }
    }
    //停止调度表
    LIN_EX_MasterStopSch(DevHandle[DevIndex],LINMasterIndex);
#endif
    //关闭设备
    USB_CloseDevice(DevHandle[DevIndex]);
	return 0;
}

